packages/flutter_tools/lib/src/runner/flutter_command.dart

```diff
@@ -1351,14 +1351,14 @@ abstract class FlutterCommand extends Command<void> {
   List<String> extractDartDefines({required Map<String, Object?> defineConfigJsonMap}) {
     final List<String> dartDefines = <String>[];
 
-    if (argParser.options.containsKey(FlutterOptions.kDartDefinesOption)) {
-      dartDefines.addAll(stringsArg(FlutterOptions.kDartDefinesOption));
-    }
-
     defineConfigJsonMap.forEach((String key, Object? value) {
       dartDefines.add('$key=$value');
     });
 
+    if (argParser.options.containsKey(FlutterOptions.kDartDefinesOption)) {
```

Comment:
can we document in the help text the precedence for these two?

Reply:
Good idea.

Reply:
Done. Also updated some error text, so that we don't yell at the user.